### PR TITLE
Dataset insertion benchmark

### PR DESCRIPTION
**Pull request by hmaarrfk**
xref: https://github.com/pydata/xarray/pull/7221

- [ ] Closes #xxxx
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: asv_bench/benchmarks/dataset_creation.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    def setup(self, elements):
+        self.datasets = {}
+        # Dictionary insertion is fast(er) than xarray.Dataser insertion
+        d = {}
+        for i in range(elements):
+            d[f"var{i}"] = i
+        self.dataset = xr.merge([d])
+
+    @parameterized(["elements"], [(0, 10, 100, 1000)])
+    def time_dataset_creation(self, elements):
```
Comment (dcherian) [inline asv_bench/benchmarks/dataset_creation.py:16]:
```suggestion
    def time_dataset_creation(self, elements):
```

File: asv_bench/benchmarks/dataset_creation.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        dataset = self.dataset
+        for i in range(5):
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply asv_bench/benchmarks/dataset_creation.py:18]: yes. thank you.

**Comment by hmaarrfk** [comment]
ok naming is always hard. I tried to pick a good name.

File: asv_bench/benchmarks/dataset_creation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply asv_bench/benchmarks/dataset_creation.py]: I think the benchmark changed as it evolved, but this is an accurate assessment of the status.

File: asv_bench/benchmarks/dataset_in_memory_operation.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import xarray as xr
+
+
+class DatasetInMemoryOperations:
+    params = [0, 10, 100, 1000]
+
+    def setup(self, elements):
+        self.datasets = {}
+        # Dictionary insertion is fast(er) than xarray.Dataser insertion
```
Comment (max-sixty) [inline asv_bench/benchmarks/dataset_in_memory_operation.py:9]:
```suggestion
        # Dictionary insertion is fast(er) than xarray.Dataset insertion
```

**Review by max-sixty — APPROVED** [review]
Excellent!

**Review by dcherian — APPROVED** [review]

File: asv_bench/benchmarks/merge.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import xarray as xr
+
+
+class DatasetAddVariable:
```
Comment (hmaarrfk) [inline asv_bench/benchmarks/merge.py:4]: TBH, i was going to expand this to other in-memory operations. Replace a variable? Drop a variable?

Comment (Illviljan) [reply asv_bench/benchmarks/merge.py:4]: I thought the first version used was relevant as well:

```python
ds = xr.Dataset(coords=coords)
for v in names:
    ds[v] = ("time", value)
```

And this fast one from https://github.com/pydata/xarray/issues/7224#issuecomment-1292216344 would also be nice to benchmark:
```python
data_vars = {v: ("time", value) for v in names}
ds = xr.Dataset(data_vars=data_vars, coords=coords)
```

Comment (hmaarrfk) [reply asv_bench/benchmarks/merge.py:4]: >  thinking that we were really timing the merge operations

It depends if you want to think about from the point of view of "xarray developer" or "xarray user".

As an "xarray user", i'm trying to "insert a totally new variable, no dims, no coords". I don't really care what is happening under the hood. This is what I originally had in mind. This is h


My goal with the original name was to test operations on in-memory xarrays (as opposed to those backed by a file, which seem to be done in `dataset_io`). I don't know if I have time to "add other benchmarks" at the moment. This seems good. 

The name is the only thing that will survive. so if you are are ok with it.

I ran the benchmarks against main, #7221, and #7222. The benchmark seems to show the effect locally even with a single insertion. I've thus simplified the benchmark to do a single insertion, and not 5.

main
```
[ 50.00%] ··· merge.DatasetAddVariable.time_variable_insertion   ok
[ 50.00%] ··· ======== ==========
               param1
              -------- ----------
                 0      636±0μs
                 10     511±0μs
                100     5.92±0ms
                1000    41.0±0ms
              ======== ==========
```

#7221
```
[ 50.00%] ··· merge.DatasetAddVariable.time_variable_insertion   ok
[ 50.00%] ··· ======== ==========
               param1
              -------- ----------
                 0      609±0μs
                 10     1.32±0ms
                100     4.01±0ms
                1000    8.65±0ms
              ======== ==========
```

#7222
```
[ 50.00%] ··· merge.DatasetAddVariable.time_variable_insertion   ok
[ 50.00%] ··· ======== ==========
               param1
              -------- ----------
                 0      578±0μs
                 10     1.03±0ms
                100     2.74±0ms
                1000    5.37±0ms
              ======== ==========
```

**Pull request merged**